### ISSUE TITLE
BUG: Cannot read property '0' of undefined

### ISSUE THREAD

**Issue by bjornharvold** [clean-css/clean-css#1042]
### Environment

* clean-css version - `4.2.0`:
* node.js version - `10.7.0`:
* operating system: Mac High Sierra

### Configuration options

```
Angular CLI 6.1.2, Angular 6.1.1
```

### Error
The error is hidden because the error reporting code fails. looks to be an invalid animation issue in my 3rd party SCSS but I am not sure what it is. I will look deeper into my CSS but line break-up.js:566 needs fixing.

asset optimization node_modules/clean-css/lib/optimizer/level-2/properties/populate-components.js:30
          throw e;
          ^

TypeError: Cannot read property '0' of undefined
    at transition (./node_modules/clean-css/lib/optimizer/level-2/break-up.js:566:93)
    at Object.breakUp (./node_modules/clean-css/lib/optimizer/level-2/break-up.js:466:28)
    at populateComponents (./node_modules/clean-css/lib/optimizer/level-2/properties/populate-components.js:17:42)
    at optimizeProperties (./node_modules/clean-css/lib/optimizer/level-2/properties/optimize.js:19:3)
    at recursivelyOptimizeProperties (/Users/crash/git/traveliko/traveliko-spa/node_modules/clean-css/lib/optimizer/level-2/optimize.js:63:9)
    at recursivelyOptimizeProperties (./node_modules/clean-css/lib/optimizer/level-2/optimize.js:66:9)

**Comment by bjornharvold** [comment]
To follow up, this is where the error is being thrown:

@include transition-default (all, 0.2s, ease);

Which is in mixins:
```css
@mixin transition-default ($what: all, $length: 0.3s, $easing: "") {
    @if ($what == transform) {
        @if ($easing != "") {
            -moz-transition: -moz-transform $length $easing;
            -o-transition: -o-transform $length $easing;
            -webkit-transition: -webkit-transform $length $easing;
            -ms-transition: -ms-transform $length $easing;
            transition: transform $length $easing;
        } @else {
            -moz-transition: -moz-transform $length;
            -o-transition: -o-transform $length;
            -webkit-transition: -webkit-transform $length;
            -ms-transition: -ms-transform $length;
            transition: transform $length;
        }
    } @else {
        @if ($easing != "") {
            -moz-transition: $what $length $easing;
            -o-transition: $what $length $easing;
            -webkit-transition: $what $length $easing;
            -ms-transition: $what $length $easing;
            transition: $what $length $easing;
        } @else {
            -moz-transition: $what $length ease-in-out;
            -o-transition: $what $length ease-in-out;
            -webkit-transition: $what $length ease-in-out;
            -ms-transition: $what $length ease-in-out;
            transition: $what $length ease-in-out;
        }
    }
}
```

Any ideas?

**Comment by bjornharvold** [comment]
FYI, this worked before v4.2.0. It stopped working once I updated Angular CLI to 6.1.2, build-ng-packager: 0.7.2 and build-angular: 0.7.2

**Comment by bjornharvold** [comment]
I can also see that the 3rd party SCSS is using this mixin in many different places with different attributes without breaking. The only difference with it breaking here that I can see is that the breaking SCSS is in an @import of the main SCSS file. Everything else is the same.

**Comment by jakubpawlowicz** [comment]
Thanks @bjornharvold for a bug report - could you also share a CSS file (as transformed SCSS) so I can narrow it down?

**Comment by bjornharvold** [comment]
![img_2045](https://user-images.githubusercontent.com/641791/43813847-b90c2940-9af1-11e8-934b-785386c19691.JPG)
![img_2046](https://user-images.githubusercontent.com/641791/43813861-d362eae0-9af1-11e8-9c92-8af4bea98902.JPG)
![img_2047](https://user-images.githubusercontent.com/641791/43813862-d39d08a6-9af1-11e8-98c8-7aec8f61497e.JPG)
![img_2048](https://user-images.githubusercontent.com/641791/43813863-d3d1011a-9af1-11e8-91bc-1199dcc5a267.JPG)
![img_2049](https://user-images.githubusercontent.com/641791/43813864-d4020ae4-9af1-11e8-8468-22f6ed662b7e.JPG)
![img_2050](https://user-images.githubusercontent.com/641791/43813866-d431dbd4-9af1-11e8-994a-c092ba1d13fd.JPG)
![img_2051](https://user-images.githubusercontent.com/641791/43813867-d463b140-9af1-11e8-8e4e-2440a2373540.JPG)

Just because I am not sure how to access this any other way :-)

**Comment by bjornharvold** [comment]
Uploaded sequentially. Sry Github rotated the pics.

**Comment by bjornharvold** [comment]
The issue is when this css is applied to the main-menu. 

SCSS all about the main menu - You can see where I commented out because clean-css failed there.
```css
@import 'styles/variables';
@import 'styles/mixin';


/* All styles for screen size over 992px
================================================== */
@media only screen and (min-width: 992px) {

	nav#menu.main-menu {
		display: block !important;
	}
	#mm-menu.main-menu {
		display: none !important;
	}
	header .btn_mobile {
		display: none !important;
	}
	/* Menu */
	.main-menu {
		position: relative;
		z-index: 9;
		width: auto;
		top: 12px;
		right: 15px;
		float: right;
		@include transition-default;
		color: #484848;
		ul,
		ul li {
			position: relative;
			display: inline-block;
			margin: 0;
			padding: 0;
			a {
				position: relative;
				margin: 0;
				padding: 0;
				display: block;
				padding: 5px 10px;
				white-space: nowrap;
			}
		}
	}
	/*First level styles */
	.main-menu>ul>li span {
		>a {
			color: #484848;
			padding: 0 8px 15px 8px;
			@include fontSize(14px);
			font-weight: 500;
		}
		&:hover>a {
			opacity: 0.7;
		}
	}
	header.sticky .main-menu>ul>li span {
		>a {
			color: #444;
		}
		&:hover>a {
			opacity: 1;
			color: $color-1;
		}
	}
	/* Submenu*/
	.main-menu ul ul {
		position: absolute;
		z-index: 1;
		visibility: hidden;
		left: 3px;
		top: 100%;
		margin: 0;
		display: block;
		padding: 0;
		background: #fff;
		min-width: 200px;
		@include box-shadow (0px 6px 12px 0px rgba(0, 0, 0, 0.175));
		transform: scale(0.4);
		transform-origin: 10% top;
		transition: 0.15s linear, 0.1s opacity cubic-bezier(0.39, 0.575, 0.565, 1), 0.15s transform cubic-bezier(0.1, 1.26, 0.83, 1);
		opacity: 0;
		//@include transition-default (all, 0.2s, ease); // TODO figure out why clean-css is breaking here
		@include border-radius(3px);
		&:before {
			bottom: 100%;
			left: 15%;
			border: solid transparent;
			content: " ";
			height: 0;
			width: 0;
			position: absolute;
			pointer-events: none;
			border-bottom-color: #fff;
			border-width: 7px;
			margin-left: -7px;
		}
		li {
			display: block;
			height: auto;
			padding: 0;
			a {
				@include fontSize(13px);
				color: $body-text-color;
				border-bottom: 1px solid #ededed;
				display: block;
				padding: 15px 10px;
				line-height: $line-height-none;
			}
		}
	}
	.main-menu ul li:hover>ul {
		padding: 0;
		visibility: visible;
		opacity: 1;
		-webkit-transform: scale(1);
		transform: scale(1);
		-webkit-transition-delay: 0.2s;
		transition-delay: 0.2s;
		-webkit-transition-duration: 0s, 0.2s, 0.2s;
		transition-duration: 0s, 0.2s, 0.2s
	}
	.main-menu ul ul li {
		&:first-child a {
			&:hover {
				@include border-radius(3px 3px 0 0);
			}
		}
		&:last-child a {
			border-bottom: none;
			&:hover {
				@include border-radius(0 0 3px 3px);
			}
		}
		&:hover>a {
			background-color: #f9f9f9;
			color: $color-1;
			padding-left: 18px;
		}
	}

}

/* All styles for screen size under 991px
================================================== */
@media only screen and (max-width: 991px) {

	nav#menu {
		display: none !important;
	}
	.mm-menu {
		background: #fff;
	}
	.main-menu {
		top: 0;
		right: 0;
		float: none;
	}
	ul.mm-listview {
		line-height: 25px;
		li {
			a {
				color: $body-text-color!important;
				display: block;
			}
		}
	}

}

```

**Comment by jakubpawlowicz** [comment]
Hmmm, that error you are getting is from Level 2 optimizations. Could you please try (in you build file) lowering the level to Level 0 and using `format: 'beautify'` option? That way you'll get a nicely formatted CSS output (instead of SCSS). If you can share that CSS file I'll be able to easily carry on from there.

**Comment by bjornharvold** [comment]
Can you say that again in Angular CLI speak ;-)

**Comment by jakubpawlowicz** [comment]
Well, I'm not familiar with Angular stack, but there has to be some sort of build file there where you can configure plugins. Is there?

**Comment by bjornharvold** [comment]
Only thing us poor folk have to fiddle with is in angular.json and that's prolly not what you need.

```
"optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "extractCss": true,
              "namedChunks": false,
              "aot": true,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true
```

**Comment by Spiderpig86** [comment]
Same issue here. It works fine if level 2 optimizations are excluded. Anyone know why that's the case? Reference: [Cirrus](https://github.com/Spiderpig86/Cirrus)

**Comment by Observablerxjs** [comment]
+1 same issue when I updated @angular-devkit/build-angular to the latest version

**Comment by Observablerxjs** [comment]
I figured out that the issue comes from one library I am using. This is the line that breaks clean-css

`@import '~bulma-extensions/dist/css/bulma-extensions.min.css';`

**Comment by lilcoder** [comment]
@bjornharvold  I got it to work when I set 
 "extractCss" to false

 "extractCss": false,

**Comment by bjornharvold** [comment]
Thank you @lilcoder :-)

**Comment by Spiderpig86** [comment]
@lilcoder Works great!
The error went away after disabling a couple of settings. (Basically set `all` to `false` and then enabling what I need).

**Comment by jakubpawlowicz** [comment]
@Spiderpig86 could you share you setup please?

**Comment by lilcoder** [comment]
@bjornharvold wahoo! :)

**Comment by Spiderpig86** [comment]
This is what I use in my gulp file:
```js
.pipe(minify({
            level: {
                1: {
                    all: true,
                    normalizeUrls: false
                  },
                2: {
                    all: false,
                    removeDuplicateRules: true,
                    reduceNonAdjacentRules: true, 
                    removeDuplicateFontRules: true, 
                    removeDuplicateMediaBlocks: true,
                    mergeAdjacentRules: true, 
                    mergeIntoShorthands: true, 
                    mergeMedia: true, 
                    mergeNonAdjacentRules: true, 
                    mergeSemantically: false, 
                    removeEmpty: true, 
                }
              }
          )
```

**Comment by jakubpawlowicz** [comment]
:+1:

**Comment by umutcakirbm** [comment]
Hi, please upgrade and fix this issue. At "node_modules\@angular-devkit\build-angular\node_modules\clean-css\lib\optimizer\level-2\break-up.js:566:93" 

`throw new InvalidPropertyError('Invalid animation value at ' + formatPosition(value[2][0]) + '. Ignoring.');`

[0] is undefined so throwing exception is also throwing exception, LOL. And I am getting this error at bootstrap.css process.

**Comment by FuelKubitox** [comment]
Any fixes for this yet? I got this Problem too. Updated everything to latest Versions, but i get this Error when i deploy my Electron Application.

**Comment by Helveg** [comment]
I've got a similar issue. 

## Versions
```
`-- @angular-devkit/build-angular@0.13.9
  `-- clean-css@4.2.1
```
## Angular error
```
 92% chunk asset optimization cleancss-webpack-pluginC:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\properties\populate-components.js:30
          throw e;
          ^

TypeError: Cannot read property '0' of undefined
    at transition (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\break-up.js:566:93)
    at Object.breakUp (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\break-up.js:466:28)
    at populateComponents (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\properties\populate-components.js:17:42)
    at optimizeProperties (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\properties\optimize.js:19:3)
    at recursivelyOptimizeProperties (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\optimize.js:63:9)
    at recursivelyOptimizeProperties (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\optimize.js:66:9)
    at level2Optimize (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\optimizer\level-2\optimize.js:77:3)
    at optimize (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\clean.js:147:5)
    at C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\clean.js:120:29
    at C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\read-sources.js:26:64
    at loadOriginalSources (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\load-original-sources.js:26:5)
    at C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\read-sources.js:26:14
    at applySourceMaps (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\apply-source-maps.js:34:5)
    at C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\read-sources.js:25:12
    at fromStyles (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\read-sources.js:152:5)
    at fromString (C:\Users\robin\Documents\GIT\angular\node_modules\clean-css\lib\reader\read-sources.js:48:10)
```

Any fix yet?

**Comment by nasiUduk** [comment]
Had the same problem because of bulma extension as well. Build angular library fix the optimization level to 2 in their CleanCssWebpackPlugins. For us this part of the css causing it :

`transition: transform .2s ease, -webkit-transform .2s ease;`

**Comment by ryanjmorris** [comment]
So this happened to me when I was trying to minify using and I'm not using an Angular stack:

`transition: 500ms flex-basis ease-in-out;`

Same happens with any flex property from as far as I could tell.

The couple fixes that I found was from @spiderpig and @lilcoder . However, I want all of the optimizations running if I can do so, and it seems to be possible. @umutcakirbm had directed me towards the correct path and there are probably multiple solutions to fix this particular issue. Without digging further and really understanding what's going on under the hood I found that adding an additional if statement in the break-up.js file in clean-css\lib\optimizer\level-2 works well:

**break-up.js in function transition(property, compactable, validator) starting on line 553**
```js
...
if (validator.isTime(value[1]) && !durationSet) {
      duration.value = [value];
      durationSet = true;
    } else if (validator.isTime(value[1]) && !delaySet) {
      delay.value = [value];
      delaySet = true;
    } else if ((validator.isGlobal(value[1]) || validator.isTimingFunction(value[1])) && !timingSet) {
      timing.value = [value];
      timingSet = true;
    } else if (validator.isIdentifier(value[1]) && !propSet) {
      prop.value = [value];
      propSet = true;
    } else {
      throw new InvalidPropertyError('Invalid animation value at ' + formatPosition(value[2][0]) + '. Ignoring.');
    }
...
```

Change this block to:

```js
...
if (validator.isTime(value[1]) && !durationSet) {
      duration.value = [value];
      durationSet = true;
    } else if (validator.isTime(value[1]) && !delaySet) {
      delay.value = [value];
      delaySet = true;
    } else if ((validator.isGlobal(value[1]) || validator.isTimingFunction(value[1])) && !timingSet) {
      timing.value = [value];
      timingSet = true;
    } else if (validator.isIdentifier(value[1]) && !propSet) {
      prop.value = [value];
      propSet = true;
    } else if (undefined === value[2]) {
      throw new InvalidPropertyError('Invalid animation value at ' + formatPosition(value[1][0]) + '. Ignoring.');
    } else {
      throw new InvalidPropertyError('Invalid animation value at ' + formatPosition(value[2][0]) + '. Ignoring.');
    }
...
```

I'm sure there is a better way to mitigate the thrown error but to always check if the value[2] is defined before throwing the InvalidPropertyError is a start. Again, without a deep dive I'm not sure if this breaks other things but then the transition property works accordingly. I'd submit a merge request but I'm thinking this isn't the best solution out there.

**Comment by CoryKleiser** [comment]
FYI by throwing different info in the library files I found that the reason I had this occurring is because we had a transition and a -webkit-transition that did not match
Something along the lines of:
`
...
 -webkit-transition: width 2s, height 4s;
 transition: width 0.5s;
...
`
removing the -webkit-transition fixed the issue. 

I found this by throwing the values in the breakup.js just after line 565. So right before `throw new InvalidPropertyError('Invalid animation value at ' + formatPosition(value[2][0]) + '. Ignoring.');` I inserted `throw values` and I found that the value `.5s` which led me to find where this issue was in the code. 
Hope this helps someone! Happy coding!
